### ISSUE TITLE
Better redirect handling for native apps?

### ISSUE THREAD

**Issue by jagthedrummer** [doorkeeper-gem/doorkeeper#1586]
### Steps to reproduce

I have a native app that registers a custom domain handler, and that uses a Doorkeeper app to handle authentication.

In the native app I click the "sign in" button and that opens a native browser window (not an in-app browser, but the actual OS browser) to a URL like this:

```
/oauth/authorize?
  client_id=xxx&
  redirect_uri=syncbot://auth-return&
  response_type=code
```

If the user is signed in the auth step happens automatically and the browser redirects to a URL like this:

```
syncbot://auth-return?code=xxx
```

Since the native app has a protocol handler registered it comes back to the foreground, can grab the code and then exchange it for a token, and then authentication is complete.

This all works well, with one wrinkle. The browser window is left open, and is just a blank window, which could be confusing for the user. (The window is blank because every URL that it has hit has just been a redirect, so nothing has ever been rendered in that window.)

If the user _wasn't_ already signed in, the initial request to `/oauth/authorize` redirects them to the sign in page, and after signing in they're redirected directly to `syncbot://auth-return`. The users still sees the login screen in their browser, even though the URL says `syncbo://auth-return`. This is also potentially confusing.

I'm wondering if it would be reasonable to make it so that if Doorkeeper sees a custom protocol in the redirect that it would first redirect to a page with some sort of messaging before redirecting to the custom protocol. I'm thinking the messaging would say something like:

> You're signed in to the web site (xyz.com) and we're now redirecting you back to the native application that requested authorization. You can close this window after you've been signed into the app.

### Expected behavior

When the `redirect_uri` is for a custom protocol Doorkeeper should render an intermediate page that tells the user that they should be getting signed in to the native app, and that they can close the window once auth is complete.

### Actual behavior

Depending on their initial sign in status the user sees either a blank screen, or they seem to remain on the sign in page.

### System configuration

N/A

### Questions

Is this something that other people would be interested in? Would a PR for this be considered for acceptance into the project?

**Comment by jagthedrummer** [comment]
After spending some time reading Doorkeeper code to figure out how/where to implement this I discovered the `response_mode=form_post` option, which already does about 98% of what I need.

On the initial authorization request I added the `response_mode` param:

```
/oauth/authorize?
  client_id=xxx&
  redirect_uri=syncbot://auth-return&
  response_type=code&
  response_mode=form_post
```

That causes Doorkeeper to render a form that is automatically `POST`ed to the redirect address.

So far I haven't figured out how to receive the `POST` body in my native (electron) app, so I need to change the method on the form to `GET`.

I ran `rails g doorkeeper:views` to extract the templates, then I edited `app/views/doorkeeper/authorizations/form_post.html.erb` to change the form `method` to `:get`. I also added some messaging to tell the user what's happening.

**Comment by jagthedrummer** [comment]
A quick follow up for anyone else who stumbles across this:

I discovered that Safari doesn't like to `POST` from an `https` url to a custom protocol. It gripes about the form posting to an insecure address.

To get around that I ended up changing `form_post.html.erb` to not post a form, but to do a JS based redirect.

After making that change everything seems to be working well in all of the major browsers.
